### ISSUE TITLE
Add support for "optional" field for proto3 files

### ISSUE THREAD

**Issue by Puneethgr** [protobuf-c/protobuf-c#476]
When there is a proto3 file which uses ```optional``` field, then protoc command throws an error.

example.proto:
```
syntax = "proto3";

message Example {
  string name = 1;
  optional int32 age = 2; // optional field is used here
}
```

## Command:
```
protoc --c_out=. example.proto
```

## Error:
```
example.proto: is a proto3 file that contains optional fields, 
but code generator protoc-gen-c hasn't been updated to support optional fields in proto3. 
Please ask the owner of this code generator to support proto3 optional.
```

## Solution:
Open ```/protobuf-c-1.4.0/protoc-c/c_generator.h``` and add the following lines inside CGenerator class:  (Refer this https://chromium.googlesource.com/external/github.com/protocolbuffers/protobuf/+/refs/heads/master/docs/implementing_proto3_presence.md)

```
  uint64_t GetSupportedFeatures() const override {
    // Indicate that this code generator supports proto3 optional fields.
    return FEATURE_PROTO3_OPTIONAL;
  }
```

## Finally, 
the CGenerator class inside ```c_generator.h``` should look like this:
```
class PROTOC_C_EXPORT CGenerator : public CodeGenerator {
 public:
  CGenerator();
  ~CGenerator();

  // implements CodeGenerator ----------------------------------------
  bool Generate(const FileDescriptor* file,
                const std::string& parameter,
                OutputDirectory* output_directory,
                std::string* error) const;
 
  uint64_t GetSupportedFeatures() const override {
    // Indicate that this code generator supports proto3 optional fields.
    return FEATURE_PROTO3_OPTIONAL;
  }

 private:
  GOOGLE_DISALLOW_EVIL_CONSTRUCTORS(CGenerator);
};
```

## At last, 
you need to re-install protobuf-c, 
go to protobuf-c-1.4.0 directory and run the installation command:
``` ./autogen.sh && ./configure && make && make install ```

Please add this support into the source code.

**Comment by acozzette** [comment]
Before updating the generator to have `GetSupportedFeatures()` return `FEATURE_PROTO3_OPTIONAL`, it is a good idea to make sure the generator has proper support for handling proto3 optional fields. Otherwise, `protoc` will run successfully but proto3 optional fields will be treated like single-element oneofs, which probably wouldn't be the ideal API. The oneof trick is a backward-compatibility measure but it's better to add proper support if possible.

**Comment by AxelLin** [comment]
Hi,
Any progress for the "optional" field support?

**Comment by devangvyas90** [comment]
Hi,
Any progress/update on this?

1. Is above advised solution good to go or will it have long term compatibility issues?
2. I am observing below issue while building following the patched solution.
In file included from protoc-c/c_generator.cc:63:0:
./protoc-c/c_generator.h:97:12: error: ‘uint64_t google::protobuf::compiler::c::CGenerator::GetSupportedFeatures() const’ marked override, but does not override
   uint64_t GetSupportedFeatures() const override {
            ^
./protoc-c/c_generator.h: In member function ‘uint64_t google::protobuf::compiler::c::CGenerator::GetSupportedFeatures() const’:
./protoc-c/c_generator.h:99:12: error: ‘FEATURE_PROTO3_OPTIONAL’ was not declared in this scope
     return FEATURE_PROTO3_OPTIONAL;
            ^
make: *** [Makefile:1609: protoc-c/protoc_c_protoc_gen_c-c_generator.o] Error 1

**Comment by samjgarnham** [comment]
Any progress?
It seems others have attempted this:
https://github.com/protobuf-c/protobuf-c/compare/master...fluent:protobuf-c:master

**Comment by AxelLin** [comment]
Hi, 
Any update?

**Comment by gsstark** [comment]
Hi, I just want to mention I ran into this problem trying to compile the opentelemetry protobufs. Specifically it seems it's the metrics protobufs that trigger this:

I assume it's this section:

```
  // min is the minimum value over (start_time, end_time].
  optional double min = 11;

  // max is the maximum value over (start_time, end_time].
  optional double max = 12;
```

**Comment by sudhank29** [comment]
> Any progress? It seems others have attempted this: [master...fluent:protobuf-c:master](https://github.com/protobuf-c/protobuf-c/compare/master...fluent:protobuf-c:master)

Incorporating this change fixes the issue of not being able to use "optional" with "proto3" syntax but the generated C structs don't have the "has_field" members.

**Comment by AxelLin** [comment]
Hi @edmonds 

Looks like v1.5.0 does not support "optional" field for proto3 files
This issue has been reported for more than 2 years, just wondering if
you have plan to fix this issue?

**Comment by luisGallet** [comment]
Any progress on adding support for optional fields for proto3?

**Comment by flachvr** [comment]
Any updates on the subject?

**Comment by Mr-Techtron** [comment]
Is there any future plan to add this feature. Its been a while.

**Comment by seven1240** [comment]
FYI. https://github.com/seven1240/protobuf-c works for me, I just updated to the latest master.

**Comment by ddomnik** [comment]
@seven1240 
> FYI. https://github.com/seven1240/protobuf-c works for me, I just updated to the latest master.

I tried your fork but the "has_" functions are not generated on my side. I am using MSYS MINGW64 to build your fork.
Could you please let me know what I did wrong?

$ protoc --version
libprotoc 31.1

**Comment by yeargun** [comment]
guys. come on now ..

**Comment by seven1240** [comment]
> [@seven1240](https://github.com/seven1240)
> 
> > FYI. https://github.com/seven1240/protobuf-c works for me, I just updated to the latest master.
> 
> I tried your fork but the "has_" functions are not generated on my side. I am using MSYS MINGW64 to build your fork. Could you please let me know what I did wrong?
> 
> $ protoc --version libprotoc 31.1

It works for me on Debian 12, I didn't try mingw. Can you try debian?

**Comment by ddomnik** [comment]
> It works for me on Debian 12, I didn't try mingw. Can you try debian?

Tried it again with WSL but no luck either.

**Comment by dkozenkov** [comment]
> FYI. https://github.com/seven1240/protobuf-c works for me, I just updated to the latest master.

"proto-gen-c" compiled from your repository only allows you to compile a proto file with optional fields.
But the *.h and *.c files compiled with your "proto-gen-c" do not contain *has_* functions for optional fields.
And this means that the original problem remains unsolved: optional fields with value containing "0" will not be packed into protobuf3, right?

**Comment by msia3dev** [comment]
Hey everyone, I've added optional fields support for one of the projects I work on, but I didn't have time to write tests for it. It has been working quite well for us for a few years. Please feel free to give it a try. It has support for `has_` parameters as well.

https://github.com/MyMila/protobuf-c/tree/feat/protobuf-3-optional-full-support

**Comment by seven1240** [comment]
> > FYI. https://github.com/seven1240/protobuf-c works for me, I just updated to the latest master.
> 
> "proto-gen-c" compiled from your repository only allows you to compile a proto file with optional fields. But the *.h and *.c files compiled with your "proto-gen-c" do not contain _has__ functions for optional fields. And this means that the original problem remains unsolved: optional fields with value containing "0" will not be packed into protobuf3, right?

Yes you are right, I just checked the generated code and seems like there's no `has_`, see if @msia3dev 's version helps.

**Comment by dkozenkov** [comment]
> Hey everyone, I've added optional fields support for one of the projects I work on, but I didn't have time to write tests for it. It has been working quite well for us for a few years. Please feel free to give it a try. It has support for `has_` parameters as well.
> 
> https://github.com/MyMila/protobuf-c/tree/feat/protobuf-3-optional-full-support

Unfortunately, your solution does not work with my complex proto file. Project compilation errors occur when the proto file contains fields of this type: "optional bytes".

**Comment by luisGallet** [comment]
We ended moving away from protobuf-c and started using nanopb instead. This
library support optional fields for proto3 files.

Luis Gallet
